File: ios/DeviceManager.swift
import Foundation
import CoreBluetooth

public class DeviceManager: NSObject, CBCentralManagerDelegate {

    private var centralManager: CBCentralManager!
    private var discoveredDevices: [Device] = []
    private var isScanning = false

    public override init() {
        super.init()
        centralManager = CBCentralManager(delegate: self, queue: nil)
    }

    public func startScan() async -> Result<[Device], PlaudError> {
        discoveredDevices.removeAll()
        isScanning = true
        centralManager.scanForPeripherals(withServices: nil, options: nil)
        
        // Simulate scan for 5 seconds
        try? await Task.sleep(nanoseconds: 5_000_000_000)
        stopScan()
        
        return .success(discoveredDevices)
    }

    public func stopScan() {
        isScanning = false
        centralManager.stopScan()
    }

    public func connect(device: Device) async -> Result<Void, PlaudError> {
        // Connection logic will be implemented
        return .success(())
    }

    public func disconnect(device: Device) async -> Result<Void, PlaudError> {
        // Disconnection logic will be implemented
        return .success(())
    }

    public func getDeviceStatus(device: Device) async -> Result<DeviceStatus, PlaudError> {
        let status = DeviceStatus(
            batteryLevel: 100,
            storageFree: 1000000,
            storageTotal: 10000000,
            isRecording: false
        )
        return .success(status)
    }

    // MARK: - CBCentralManagerDelegate

    public func centralManagerDidUpdateState(_ central: CBCentralManager) {
        // Handle Bluetooth state updates
    }

    public func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String : Any], rssi RSSI: NSNumber) {
        let device = Device(
            name: peripheral.name,
            address: peripheral.identifier,
            rssi: RSSI.intValue,
            connectionState: .disconnected
        )
        
        if !discoveredDevices.contains(where: { $0.address == device.address }) {
            discoveredDevices.append(device)
        }
    }

    public func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        // Handle successful connection
    }
}

extension Result {
    public static func success(_ value: Success) -> Result<Success, Failure> {
        return .success(value)
    }

    public static func failure(_ error: Failure) -> Result<Success, Failure> {
        return .failure(error)
    }
}

File: android/src/main/java/com/memotalk/sdk/DeviceManager.kt
package com.memotalk.sdk

import android.bluetooth.*
import android.content.Context
import kotlinx.coroutines.*
import java.util.concurrent.CopyOnWriteArrayList

class DeviceManager(private val context: Context) {

    private val bluetoothAdapter: BluetoothAdapter? by lazy {
        val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
        bluetoothManager.adapter
    }

    private val scanner = BluetoothLeScanner()
    private val discoveredDevices = CopyOnWriteArrayList<Device>()
    private var isScanning = false

    private val scanCallback = object : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult?) {
            super.onScanResult(callbackType, result)
            result?.let { scanResult ->
                val device = Device(
                    name = scanResult.device.name,
                    address = scanResult.device.address,
                    rssi = scanResult.rssi,
                    nativeDevice = scanResult.device
                )
                if (!discoveredDevices.any { it.address == device.address }) {
                    discoveredDevices.add(device)
                }
            }
        }

        override fun onScanFailed(errorCode: Int) {
            super.onScanFailed(errorCode)
            // Handle scan failure
        }
    }

    suspend fun startScan(): Result<List<Device>, PlaudError> = withContext(Dispatchers.Default) {
        return@withContext try {
            if (bluetoothAdapter == null) {
                Result.failure(PlaudError(1001, "Bluetooth not supported"))
            } else {
                discoveredDevices.clear()
                isScanning = true
                scanner.startScan(scanCallback)
                Result.success(discoveredDevices.toList())
            }
        } catch (e: Exception) {
            Result.failure(PlaudError(1002, "Scan failed: ${e.message}"))
        }
    }

    suspend fun stopScan(): Result<Unit, PlaudError> = withContext(Dispatchers.Default) {
        return@withContext try {
            isScanning = false
            scanner.stopScan(scanCallback)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(PlaudError(1003, "Stop scan failed: ${e.message}"))
        }
    }

    suspend fun connect(device: Device): Result<Unit, PlaudError> = withContext(Dispatchers.IO) {
        return@withContext try {
            device.connectionState = ConnectionState.CONNECTING
            // Connection logic will be implemented
            device.connectionState = ConnectionState.CONNECTED
            Result.success(Unit)
        } catch (e: Exception) {
            device.connectionState = ConnectionState.DISCONNECTED
            Result.failure(PlaudError(2001, "Connection failed: ${e.message}"))
        }
    }

    suspend fun disconnect(device: Device): Result<Unit, PlaudError> = withContext(Dispatchers.IO) {
        return@withContext try {
            device.connectionState = ConnectionState.DISCONNECTING
            // Disconnection logic will be implemented
            device.connectionState = ConnectionState.DISCONNECTED
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(PlaudError(2002, "Disconnection failed: ${e.message}"))
        }
    }

    suspend fun getDeviceStatus(device: Device): Result<DeviceStatus, PlaudError> = withContext(Dispatchers.IO) {
        return@withContext try {
            // Status query logic will be implemented
            Result.success(DeviceStatus(
                batteryLevel = 100,
                storageFree = 1000000,
                storageTotal = 10000000,
                isRecording = false
            ))
        } catch (e: Exception) {
            Result.failure(PlaudError(3001, "Status query failed: ${e.message}"))
        }
    }
}

sealed class Result<out S, out E> {
    data class Success<S>(val value: S) : Result<S, Nothing>()
    data class Failure<E>(val error: E) : Result<Nothing, E>()

    companion object {
        fun <S> success(value: S): Result<S, Nothing> = Success(value)
        fun <E> failure(error: E): Result<Nothing, E> = Failure(error)
    }

    fun <T> map(transform: (S) -> T): Result<T, E> = when (this) {
        is Success -> Success(transform(value))
        is Failure -> Failure(error)
    }

    fun <T> flatMap(transform: (S) -> Result<T, E>): Result<T, E> = when (this) {
        is Success -> transform(value)
        is Failure -> Failure(error)
    }
}
